packages/flutter_tools/lib/src/base/error_handling_io.dart

```diff
@@ -584,15 +650,88 @@ class ErrorHandlingProcessManager extends ProcessManager {
     Encoding stdoutEncoding = io.systemEncoding,
     Encoding stderrEncoding = io.systemEncoding,
   }) {
-    return _runSync(() => _delegate.runSync(
-      command,
-      workingDirectory: workingDirectory,
-      environment: environment,
-      includeParentEnvironment: includeParentEnvironment,
-      runInShell: runInShell,
-      stdoutEncoding: stdoutEncoding,
-      stderrEncoding: stderrEncoding,
-    ), platform: _platform);
+    return _runSync(() {
+      return _processRunSync(
+        _getCommandPath(command, workingDirectory),
+        _getArguments(command),
+        workingDirectory: workingDirectory,
+        environment: environment,
+        includeParentEnvironment: includeParentEnvironment,
+        runInShell: runInShell,
+        stdoutEncoding: stdoutEncoding,
+        stderrEncoding: stderrEncoding,
+      );
+    }, platform: _platform);
+  }
+
+  /// Resolve the first item of [command] to a file or link.
+  ///
+  /// This function will cache each raw command string per working directory,
+  /// so that repeated lookups for the same command will resolve to the same
+  /// executable. If a new executable is installed in a higher priority
+  /// location after a command has already been cached, the older executable
+  /// will be used until the tool restarts.
+  ///
+  /// If the entity cannot be found, return the raw command as is and do not
+  /// cache it.
+  String _getCommandPath(List<dynamic> command, String workingDirectory, { bool strict = false }) {
+    final String executable = command.first.toString();
+    final Map<String, String> workingDirectoryCache = _resolvedExecutables[workingDirectory] ??= <String, String>{};
+    final String resolvedExecutable = workingDirectoryCache[executable]
+      ?? _which(executable, workingDirectory);
+    if (resolvedExecutable != executable) {
+      workingDirectoryCache[executable] = resolvedExecutable;
+    }
+    return resolvedExecutable;
+  }
+
+  String _which(String execName, String workingDirectory) {
```

Comment:
Can this be unified with, or can this be a wrapper around, the code in os.dart?

Reply:
I think I could refactor OS to use `which` exposed through the process manager class (once our dep on package:process is removed), but I can't do that before this is landed since that creates a circular dependency